### PR TITLE
#23 - Allow nullable elements in a collection

### PR DESCRIPTION
**Pull request by JesusValera**
## 📚 Description

[Issue ticket](https://github.com/Chemaclass/typed-arrays/issues/23)

Allow creating collections that can hold `nullable` values.

## 🔖 Changes

- Improve documentation with a new `Nullability` category
- Refactorize the `example/articles.php` with `null` value as an element in the collection
- Implement the `isNullAllowed()` method and the logic at `AbstractTypedArray`
- Some tests

### CODE REVIEW

File: src/AbstractTypedArray.php [hunk 1]
```diff
@@ -147,6 +152,10 @@ private function guardInstanceTypeToEnforce(array $input): void
      */
     private function checkType($variable): bool
     {
+        if (is_null($variable)) {
```
Comment (Chemaclass) [inline src/AbstractTypedArray.php:155]: I preffer `null === variable`. Why? Because -> https://stackoverflow.com/a/8229005/3454593

**Review by Chemaclass — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -108,14 +113,26 @@ protected function collectionType(): string
...
 
+## Nullability
+
+A collection can allow `null` items or not. By default, the collections cannot allow `null`s.
+
+You can specify the collection allows nullable items or not by overriding this function in your domain:
+```php
+protected function isNullAllowed(): bool
+{
+    return false;
```
Comment (Tito-Kati) [inline README.md:132]: If you're explaining how to make the list nullable, shouldn't the example be `true`?

File: README.md [hunk 3]
```diff
@@ -153,13 +170,13 @@ If you only need a string map, or a list of integers, this library already conta
 
 You can check some working examples in `example` folder
 ```bash
-# *ImmutableArticleList* wrapper from Article instances
+# The `ImmutableArticleList` is a wrapper of Article instances
```
Comment (Tito-Kati) [inline README.md:173]: It is? Is it a wrapper? Rather, it is a list of articles, right?

File: example/articles.php [hunk 4]
```diff
@@ -54,16 +59,18 @@ protected function collectionType(): string
 function renderArticles(ImmutableArticleList $articles): void
 {
     foreach ($articles as $article) {
-        echo $article . PHP_EOL;
+        echo($article ?? 'NULL') . PHP_EOL;
```
Comment (Tito-Kati) [inline example/articles.php:62]: I'm not sure if this example is the best to demonstrate this functionality, what if you create a new one?

So thinking quickly, you could make one with 2 translation maps (i18n you know) in which you use English as a fallback to a null in the other language.

So incidentally we have an example with maps that we do not have any.

File: tests/Unit/NullabilityTest.php [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace TypedArraysTest\Unit;
+
+use PHPUnit\Framework\TestCase;
+use TypedArrays\Exceptions\InvalidTypeException;
+use TypedArraysTest\Unit\Fixtures\NonNullableSimpleObjectArray;
+use TypedArraysTest\Unit\Fixtures\NullableSimpleObjectArray;
+use TypedArraysTest\Unit\Fixtures\SimpleObject;
+
+final class NullabilityTest extends TestCase
```
Comment (Tito-Kati) [inline tests/Unit/NullabilityTest.php:13]: I think there are some tests missing here for the setters.
```php
$test[] = null;
```
This should work also if the array is nullable and it shouldn't work if it isn't.

**Review by Tito-Kati — CHANGES_REQUESTED** [review]

**Comment by Tito-Kati** [comment]
What are we going to do with the default nullable scalars? @Chemaclass @JesusValera

**Comment by Chemaclass** [comment]
> What are we going to do with the default nullable scalars? @Chemaclass @JesusValera

I wouldn’t do this for now... We should think about this in a future iteration.

**Pull request merged**